data/lua/core/unit_test.lua

```diff
@@ -0,0 +1,204 @@
+
+--[========[Unit Test Helpers]========]
+
+if rawget(_G, 'unit_test') ~= nil then
+	print("Loading unit_test module...")
+	
+	--! Causes the test to complete with a specific result.
+	--! This is mostly an internal detail but might very occasionally be useful to produce a result other than pass or fail
+	function unit_test.finish(result)
+		wesnoth.wml_actions.endlevel{
+			test_result = result,
+			linger_mode = true
+		}
+	end
+
+	--! Fail the test unconditionally
+	function unit_test.fail()
+		unit_test.finish('fail')
+	end
+
+	--! Pass the test unconditionally
```

Comment:
What happens if neither `pass()` nor `fail()` is called? Will it be like `{SUCCEED}`, where the test ends up running endlessly?

What happens if an assert fails first? Does Lua execution skip the `pass()`, does the `pass()` override the earlier status, or does `pass()` conditionally not cause the test to pass?
```
unit_test.assert(x == 1, "Wrong value for x")
unit_test.pass()
```

Reply:
If neither is called, then the test will never terminate.

These use `[endlevel]` under the hood, and as I understand it, the first `[endlevel]` wins and any subsequent ones do nothing. Thus, an assert failing means a subsequent pass does nothing.

Reply:
The documentation should reflect that.

Keeping compatibility with the quirks of the WML unit test framework might help when we replace that (I assume we'll port the old macros to be based on the Lua unit test framework), but a new implementation should be an opportunity to change to a better-understood implementation.

Reply:
I'm not sure why the WML unit test framework would be going away?  While many WML tags are implemented in lua under the hood, WML is still its own API that needs its own tests.

Reply:
This comment is a wishlist, not a request for the current PR.

> I'm not sure why the WML unit test framework would be going away?

The WML unit test framework has a mechanism (the `ended` flag) to stop logging asserts after an assert has already failed. It would be nice to add similar functionality to the Lua unit test framework. It would then be nice to have the two unit test frameworks interact with each other, so that logging stops on the first failed assert even when a test mixes the two frameworks. Conclusion: in 1.18 I expect the API (well the macro names) of the WML unit test framework to still be there, but for those macros to have become a wrapper around the Lua unit test framework.

Reply:
That `ended` flag seems like something that's better left optional… there might be some asserts where you want to keep going (because if multiple asserts fail you can fix them all at once) and others where you don't (because the assertion failure guarantees that later asserts also fail).

Reply:
Marking this conversation as resolved - the initial point about more information is still open but is also covered by https://github.com/wesnoth/wesnoth/pull/5708#discussion_r616272336